bisync: more fixes for integration tests

nielash opened the pull request:
#### What is the purpose of this change?

More fixes for integration tests:

- use `fstest.RandomRemote` to create the root level directory
- fix parsing of canonical name for connection string remotes
- `hdfs`: fix `f.String()` not including subpath
- `webdav`: fix `SetModTime` erasing checksums on `owncloud` and `nextcloud`

#### Was the change discussed in an issue or in the forum before?

- #7665

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

nielash commented:
I think this is failing because there's a problem with the test on https://github.com/rclone/rclone/commit/e053c8a1c03e9c4396e3588bbfc432ed4ff79814 (sorry -- didn't realize you'd be merging that so soon!)

I re-ran the CI in my fork and it is passing: https://github.com/nielash/rclone/actions/runs/8525554442

ncw reviewed (approved):
These look good to me - please merge.

The webdav fix is a good one.

ncw commented:
@nielash the tests are looking better! The major failures seem to be timeouts now.

Taking a specific example https://pub.rclone.org/integration-tests/current/s3-cmd.bisync-TestS3MinioEdge-5.txt this is running against a local S3 server so should be quick.

However we get a 60 minute timeout.

Looking at the log

```
"./bisync.test -test.v -test.timeout 1h0m0s -remote TestS3MinioEdge: -verbose" - Starting (try 5/5)
2024/04/04 05:42:19 DEBUG : Creating backend with remote "TestS3MinioEdge:rclone-test-toyotec8betover2sodepob6"
...
2024/04/04 05:42:35 INFO  : ě�� áñhࢺ_測試рускийěáñ👸🏼🧝🏾‍♀️💆🏿‍♂️🐨🤙🏼🤮🧑🏻‍🔧🧑‍🔬éö: Copied (new)
panic: test timed out after 1h0m0s
```

It seems to have got stuck after running for 16 seconds and I can confirm this when running the tests locally against a local minio.

There are >1200 go routines active which is way too many!

As far as I can see (using panic parse) the only runnable bit of code is this

```
1: runnable [Created by testing.(*T).Run in goroutine 52314 @ testing.go:1648]
    json        stream.go:49          (*Decoder).Decode(*Decoder(#9921), #48, #67)
    bisync      lockfile.go:102       (*bisyncRun).lockFileIsExpired(*bisyncRun(#138))
    bisync      lockfile.go:34        (*bisyncRun).setLockFile(*bisyncRun(#138))
    bisync      operations.go:122     Bisync(Context{#59, #7833}, Fs{#64, #7933}, Fs{#64, #9007}, *Options(#10806))
    bisync_test bisync_test.go:1122   (*bisyncTest).runBisync.func3()
    bilib       output.go:18          CaptureOutput(func(#10805))
    bisync_test bisync_test.go:1121   (*bisyncTest).runBisync(*bisyncTest(#9425), Context{#57, #71}, []string(#6485 len=2 cap=0))
    bisync_test bisync_test.go:726    (*bisyncTest).runTestStep(*bisyncTest(#9425), Context{#57, #71}, string(#9987, len=62))
    bisync_test bisync_test.go:473    (*bisyncTest).runTestCase(*bisyncTest(#9425), Context{#57, #71}, *T(#65), string(#9303, len=824657137504))
    bisync_test bisync_test.go:344    testBisync.func4(*T(0x0))
    testing     testing.go:1595       tRunner(#10614, #9583)
    testing     testing.go:1648       (*T).Run.func1()
    runtime     asm_amd64.s:1650      goexit()
```

So the culprit is probably in there!

nielash commented:
Hmm that's odd! I'll try to debug it.

ncw commented:
There is a hint [here](https://github.com/rclone/rclone/blob/master/fstest/testserver/init.d/TestS3Minio) on how to run a local Minio server with docker if you need one.

nielash commented:
> There is a hint [here](https://github.com/rclone/rclone/blob/master/fstest/testserver/init.d/TestS3Minio) on how to run a local Minio server with docker if you need one.

I discovered that recently! FYI I've been finding they only work for me if I add a custom port mapping with `-p`... maybe a mac-specific thing?

ncw commented:
Yes it's a Mac thing. Someone else complained about that. It's because on Mac the docker images run in a VM whereas on Linux they run more natively in a container. 

This could be fixed though, we'd just need to make sure we assign a unique port number to each one. Some of them have been converted already.

ncw commented:
@nielash 

I had a go at fixing the integration tests here: [v1.67.0-beta.7853.00737f1e7.fix-macos-integration-tests](https://beta.rclone.org/branch/fix-macos-integration-tests/v1.67.0-beta.7853.00737f1e7.fix-macos-integration-tests/) on branch [fix-macos-integration-tests](https://github.com/rclone/rclone/tree/fix-macos-integration-tests) (uploaded in 15-30 mins)

nielash commented:
Excellent! This will be very helpful for debugging 🙂 

While we're on the topic, thought I'd mention two other little issues I stumbled over at first (I figured out workarounds eventually):

- `make test` doesn't seem to put the `test_all` binary in the right place on mac
- the `-remotes` flag doesn't play nice with connection string remotes like `TestS3,directory_markers:` -- I think because it splits the string with a comma separator [here](https://github.com/rclone/rclone/blob/46b385433082b105bcdf10f3e9ca94789db423ef/fstest/test_all/test_all.go#L81).

ncw commented:
> Excellent! This will be very helpful for debugging 🙂

If you could test it works on macOS that would be super helpful.  Something like `test_all -verbose -remotes TestS3Minio: -tests backend` should work.

> While we're on the topic, thought I'd mention two other little issues I stumbled over at first (I figured out workarounds eventually):
> 
> * `make test` doesn't seem to put the `test_all` binary in the right place on mac

Do you have `go env GOPATH` on your PATH? That then makes `go install` work.

> * the `-remotes` flag doesn't play nice with connection string remotes like `TestS3,directory_markers:` -- I think because it splits the string with a comma separator [here](https://github.com/rclone/rclone/blob/46b385433082b105bcdf10f3e9ca94789db423ef/fstest/test_all/test_all.go#L81).

Hmm, yes, I hadn't noticed that. It should probably use the CSV parser to undo the `,` then that gives opportunity for quoting.

nielash commented:
> If you could test it works on macOS that would be super helpful.

I tried this:

```
test_all -remotes TestS3Minio:,TestS3MinioEdge:,TestSFTPOpenssh:,TestSMB:rclone,TestSwiftAIO:,TestWebdavNextcloud: -tests backend -timeout 1h -verbose -maxtries 1 -n 2
```

(btw, might be worth documenting somewhere that it needs to be `TestSMB:rclone` not `TestSMB:`)

These passed:

- `TestSMB:rclone`

These failed some of their tests, but oddly ended up in the "Passed" section of the report anyway:

- `TestS3Minio:`
- `TestS3MinioEdge:`
- `TestWebdavNextcloud:`

These failed and were reported as such:

- `TestSFTPOpenssh:`

```
TestSFTPOpenssh: Connection to "127.0.0.1:28627" failed try 30/30: dial tcp 127.0.0.1:28627: connect: connection refused
```

- `TestSwiftAIO:`

```
Error:      	Received unexpected error:
        	            	Get "http://127.0.0.1:28628/auth/v1.0": EOF
        	Test:       	TestIntegration
        	Messages:   	unexpected error: Get "http://127.0.0.1:28628/auth/v1.0": EOF
```



> Do you have `go env GOPATH` on your PATH? That then makes `go install` work.

That's probably the issue!

> Hmm, yes, I hadn't noticed that. It should probably use the CSV parser to undo the , then that gives opportunity for quoting.

Here's the very quick-and-dirty workaround hack I've been using 😆 

```diff --git a/fstest/test_all/test_all.go b/fstest/test_all/test_all.go
index 79263f0e4..c78e02e64 100644
--- a/fstest/test_all/test_all.go
+++ b/fstest/test_all/test_all.go
@@ -78,7 +78,15 @@ func main() {
 
 	// Filter selection
 	if *testRemotes != "" {
-		conf.filterBackendsByRemotes(strings.Split(*testRemotes, ","))
+		const comma = "REPLACE_ME_WITH_COMMA" // allow connection string remotes if comma escaped
+		remotes := strings.ReplaceAll(*testRemotes, `\,`, comma)
+		splitRemotes := strings.Split(remotes, ",")
+		for i, r := range splitRemotes {
+			if strings.Contains(r, comma) {
+				splitRemotes[i] = strings.ReplaceAll(r, comma, ",")
+			}
+		}
+		conf.filterBackendsByRemotes(splitRemotes)
 	}
 	if *testBackends != "" {
 		conf.filterBackendsByBackends(strings.Split(*testBackends, ","))
```

ncw commented:
Ah, I knew I'd find this comment eventually!

> > If you could test it works on macOS that would be super helpful.
> 
> I tried this:
> 
> ```
> test_all -remotes TestS3Minio:,TestS3MinioEdge:,TestSFTPOpenssh:,TestSMB:rclone,TestSwiftAIO:,TestWebdavNextcloud: -tests backend -timeout 1h -verbose -maxtries 1 -n 2
> ```
> 
> (btw, might be worth documenting somewhere that it needs to be `TestSMB:rclone` not `TestSMB:`)

I think it says that in [the backend docs](https://rclone.org/smb/) but it caught me out too!

> These passed:
> 
> * `TestSMB:rclone`

Good.

> These failed some of their tests, but oddly ended up in the "Passed" section of the report anyway:
> 
> * `TestS3Minio:`
> * `TestS3MinioEdge:`
> * `TestWebdavNextcloud:`

I think you've seen the reason for that in `config.yaml` now - failing tests we can `ignore`.

> These failed and were reported as such:
> 
> * `TestSFTPOpenssh:`
> 
> ```
> TestSFTPOpenssh: Connection to "127.0.0.1:28627" failed try 30/30: dial tcp 127.0.0.1:28627: connect: connection refused
> ```
> 
> * `TestSwiftAIO:`
> 
> ```
> Error:      	Received unexpected error:
>         	            	Get "http://127.0.0.1:28628/auth/v1.0": EOF
>         	Test:       	TestIntegration
>         	Messages:   	unexpected error: Get "http://127.0.0.1:28628/auth/v1.0": EOF
> ```

Thes both look like connection errors - the paths were correct...

If you do this in that branch

```
 ./fstest/testserver/init.d/TestSwiftAIO start
b68ceed46d2c72a1eb23dee1bad9bf868c0472cdec7f8b5042150b7052b7fe66
type=swift
env_auth=false
user=test:tester
key=testing
auth=http://127.0.0.1:28628/auth/v1.0
_connect=127.0.0.1:28628
```

Then

```
$ curl http://127.0.0.1:28628/auth/v1.0
<html><h1>Unauthorized</h1><p>This server could not verify that you are authorized to access the document you requested.</p></html>
```

You should get that nice blob of HTML

> > Do you have `go env GOPATH` on your PATH? That then makes `go install` work.
> 
> That's probably the issue!
> 
> > Hmm, yes, I hadn't noticed that. It should probably use the CSV parser to undo the , then that gives opportunity for quoting.
> 
> Here's the very quick-and-dirty workaround hack I've been using 😆

Eww!

nielash commented:
> ```
> $ curl http://127.0.0.1:28628/auth/v1.0
> <html><h1>Unauthorized</h1><p>This server could not verify that you are authorized to access the document you requested.</p></html>
> ```
> 
> 
>     
>   
> 
> You should get that nice blob of HTML

Indeed that's exactly what I get for `TestSwiftAIO`.

For `TestSFTPOpenssh` I get:

```
% curl http://127.0.0.1:28627/auth/v1.0          
curl: (7) Failed to connect to 127.0.0.1 port 28627 after 0 ms: Couldn't connect to server
```

It looks to me like the docker container starts running but then immediately stops (as opposed to `TestSwiftAIO` which keeps running).

Also FYI, because I'm on an Apple Silicon machine, docker gives me an `AMD64` warning for some containers: "Image may have poor performance, or fail, if run via emulation".

ncw commented:
Looks like the swift one is working then???

For the open ssh one

```
$ ./fstest/testserver/init.d/TestSFTPOpenssh start
03da80df8fd82508da57b5fd85730795a95f931fae27961d9cb4021b24eec7b8
type=sftp
host=127.0.0.1
port=28627
user=rclone
pass=gTDzJYBXMKNfI1fI89hYYHtZhh1Yd72c
copy_is_hardlink=true
_connect=127.0.0.1:28627
```

You should get this

```
$ telnet 127.0.0.1 28627
Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.
SSH-2.0-OpenSSH_8.1
^]
telnet> close
Connection closed.
```

I think I'll merge that branch as it is an improvement anyway.

nielash commented:
Instead I get:

```
% telnet 127.0.0.1 28627                          
Trying 127.0.0.1...
telnet: connect to address 127.0.0.1: Connection refused
telnet: Unable to connect to remote host
```

Which I think is because the container is no longer running. (It stops basically as soon as it's started)

(Btw, `telnet` is not installed by default on mac, but I doubt that matters.)

> Looks like the swift one is working then???

I can get it to work by removing `-n 2`. I guess it just needs some time -- I get this about 6 times:

```
2024/04/16 15:12:56 DEBUG : TestSwiftAIO: Read 0, error: EOF
2024/04/16 15:12:57 DEBUG : TestSwiftAIO: Attempting to connect to "127.0.0.1:28628" try 6/30
```

And then it works the 7th time. `¯\_(ツ)_/¯`
